packages/puppeteer-core/src/common/IterableUtil.ts

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright 2023 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type {JSHandle} from '../api/JSHandle.js';
+import type {AwaitableIterable, HandleFor} from './types.js';
+
+/**
+ * @internal
+ */
+export class IterableUtil {
+  static async *map<T, U>(
+    iterable: AwaitableIterable<T>,
+    map: (item: T) => Promise<U>
+  ): AwaitableIterable<U> {
+    for await (const value of iterable) {
+      yield await map(value);
+    }
+  }
+
+  static async collect<T>(iterable: AwaitableIterable<T>): Promise<T[]> {
+    const result = [];
+    for await (const value of iterable) {
+      result.push(value);
+    }
+    return result;
+  }
+
+  static async first<T>(
+    iterable: AwaitableIterable<T>
+  ): Promise<T | undefined> {
+    for await (const value of iterable) {
+      return value;
+    }
+    return undefined;
+  }
+
+  static async *transposeHandle<T>(
+    handle: JSHandle<AwaitableIterable<T>>
+  ): AwaitableIterable<HandleFor<T>> {
+    const iterator = await handle.evaluateHandle(iterable => {
+      return (async function* () {
+        yield* iterable;
+      })();
+    });
+    try {
+      while (true) {
+        const handle = await iterator.evaluateHandle(async iterator => {
```

Comment:
this looks expensive because a CDP call is made of every iteration. Did we do a separate call for every returned element in the previous version? I see that this method is used for all query handlers now.

Reply:
Also, if I understand it right: with this change the querySelectorAll is not atomic anymore because changes can happen while the iterator is being invoked. I am not sure how big of the effect it might have but I wonder if we can check the following things:

1) benchmark a query all with the old way and new way (time-wise)
2) see if we can have the collectAll to be executed in the page context

Reply:
> with this change the querySelectorAll is not atomic anymore because changes can happen while the iterator is being invoked

We can't use `collect` in the page context because in the upcoming `P` queries PR, the iterators may not be based on the page context.

> benchmark a query all with the old way and new way (time-wise)

We can actually test this easily surprisingly. We have a test (`should handle many elements`) for excess elements which gives us a time estimate:  For a simple CSS selector of 1000 elements, we have

 - 597ms using iterators,
 - 188ms without it.

However I have an assumption that no one every queries that many elements repetitively. Also, with the iterator approach, we don't have to use `getProperties` which is a CDP implementation.

Reply:
If you are wondering if it's linear growth, it is. For DevTools, let me see.

Reply:
https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/4241082

Reply:
I've implemented batching. With batching, the speed is about 300ms which is quite reasonable. I've set the batch sizes to 20 for now.